Utils/XDrive.hpp

```diff
@@ -57,8 +57,130 @@ class XDriveRobot {
+  // tgtHeading is a parameter
+  double targetArc = degree2arc(tgtHeading);
+  
+  // tracking distance
+  double distToGo = dist; // distance more to travel
+  double distTravelled = dist - distToGo; // distance already traveled
+  double prevRot = backleftdrive.rotation(vex::rotationUnits::deg); // amount of rotation at the last run through
+  double changedRot = backleftdrive.rotation(vex::rotationUnits::deg) - prevRot; // rotations passed since last loop
+
+  // speed
+  double finalSpeed = 10; // capping speed
+  double speed = originalSpeed; // max speed
+  double const adaptiveInterval = 10; // slow down/speed up interval
+
+  
+  // location tracking on the coordinate plane
+  double dx = 0; // change in x coordinate since last loop
+  double dy = 0; // change in y coordinate since last loop
+
+  double xSpeed = originalSpeed * cos(targetArc);
+  double ySpeed = originalSpeed * sin(targetArc);
+
+  double blspeed = xSpeed - ySpeed;
+  double brspeed = -xSpeed - ySpeed;
+  double flspeed = -xSpeed + ySpeed;
+  double frspeed = xSpeed + ySpeed;
+
+  while (distToGo > 0) {
+
+    double frsign = 1;
+    double flsign = -1;
+    double blsign = -1;
+    double brsign = 1;
+
+    // update other variables
+
+    double headingError = inertialSensor.heading() - tgtHeading; // updates the error in heading
+
+    // adjusts the heading error so that it is between -180 and 180
+    if (headingError < -180) headingError = headingError + 360;
+    if (headingError > 180) headingError = headingError - 360;
+
+    // caps headingError so that adjustments are not too dramatic
+    if (headingError < -15) headingError = -15;
+    if (headingError > 15) headingError = 15;
+
+    // by default, speed with no adjustments is as such
+    speed = originalSpeed;
+
+    if (distTravelled < adaptiveInterval) speed = originalSpeed * distTravelled / adaptiveInterval;
+    if (distToGo < adaptiveInterval) speed = originalSpeed * (1 - (adaptiveInterval - distToGo) / adaptiveInterval);
+
+    if (speed < finalSpeed) speed = finalSpeed;
+
+    if (headingError > 0) {
+        move(blspeed * (1 + kp * headingError * blsign), brspeed * (1 + kp * headingError * brsign), flspeed * (1 + kp * headingError * flsign), frspeed * (1 + kp * headingError * frsign));
+    }
+
+    else {
+        move(blspeed * (1 - kp * headingError * blsign), brspeed * (1 - kp * headingError * brsign), flspeed * (1 - kp * headingError * flsign), frspeed * (1 - kp * headingError * frsign));
+    }
+
+    vex::task::sleep(10);
+
+    distToGo = dist - fabs(leftdrive.rotation(vex::rotationUnits::deg) / 360 * (4.0 * 3.1415269265));
+
+    changedRot = (leftdrive.rotation(vex::rotationUnits::deg) * 4.0 * 3.1415269265) / 360 - prevRot; // need scale, deg => inch
+
+    dx = changedRot * cos(90-inertialSensor.heading());
+    dy = changedRot * sin(90-inertialSensor.heading());
+
+    prevRot = (leftdrive.rotation(vex::rotationUnits::deg) * 4.0 * 3.1415269265) / 360;
```

Comment:
create a new local variable (within this function) called refDrive (the driver as reference to measure distance). For now, you can assign it to backleftdrive. We can revisit such choice later if the accuracy is not enough.